Bluetooth: Tester: Use BT_L2CAP_SEG_RECV for L2CAP tests

sjanc opened the pull request:
This also fixes some issues I've found while using segment direct API.
I also added some TODO comments that could be a starting point for small API improvements

sjanc committed:
Bluetooth: Host: L2CAP: Fix seg_recv call on SDU overflow

This fix calling seg_recv() callback being called even though channel
is being disconnected due to SDU overflow.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
Bluetooth: Tester: Use BT_L2CAP_SEG_RECV for L2CAP tests

This API gives better control on L2CAP COC credits and suits better
for Upper Tester implementation.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
Bluetooth: Host: L2CAP: Fix reconfigure with Segment Direct API

When Segment Direct API is used MPS is handled by application and
thus host shall obey that.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
#AutoPTS run zephyr L2CAP

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/80302#issuecomment-2432070110, board: nrf52, estimated start time: 15:00:27, test case count: 93, estimated duration: 0:24:54<details><summary>Test cases to be run</summary>L2CAP/COS/CFC/BV-01-C<br>L2CAP/COS/CFC/BV-02-C<br>L2CAP/COS/CFC/BV-03-C<br>L2CAP/COS/CFC/BV-04-C<br>L2CAP/COS/CFC/BV-05-C<br>L2CAP/COS/ECFC/BV-01-C<br>L2CAP/COS/ECFC/BV-02-C<br>L2CAP/COS/ECFC/BV-03-C<br>L2CAP/COS/ECFC/BV-04-C<br>L2CAP/LE/CPU/BV-01-C<br>L2CAP/LE/CPU/BV-02-C<br>L2CAP/LE/CPU/BI-01-C<br>L2CAP/LE/CPU/BI-02-C<br>L2CAP/LE/REJ/BI-02-C<br>L2CAP/LE/CFC/BV-01-C<br>L2CAP/LE/CFC/BV-02-C<br>L2CAP/LE/CFC/BV-03-C<br>L2CAP/LE/CFC/BV-04-C<br>L2CAP/LE/CFC/BV-05-C<br>L2CAP/LE/CFC/BV-06-C<br>L2CAP/LE/CFC/BV-07-C<br>L2CAP/LE/CFC/BI-01-C<br>L2CAP/LE/CFC/BV-08-C<br>L2CAP/LE/CFC/BV-09-C<br>L2CAP/LE/CFC/BV-10-C<br>L2CAP/LE/CFC/BV-11-C<br>L2CAP/LE/CFC/BV-12-C<br>L2CAP/LE/CFC/BV-13-C<br>L2CAP/LE/CFC/BV-14-C<br>L2CAP/LE/CFC/BV-15-C<br>L2CAP/LE/CFC/BV-16-C<br>L2CAP/LE/CFC/BV-18-C<br>L2CAP/LE/CFC/BV-19-C<br>L2CAP/LE/CFC/BV-20-C<br>L2CAP/LE/CFC/BV-21-C<br>L2CAP/LE/CFC/BV-22-C<br>L2CAP/LE/CFC/BV-23-C<br>L2CAP/LE/CFC/BV-24-C<br>L2CAP/LE/CFC/BV-25-C<br>L2CAP/LE/CFC/BV-26-C<br>L2CAP/LE/CFC/BV-27-C<br>L2CAP/LE/CFC/BV-28-C<br>L2CAP/LE/CFC/BV-29-C<br>L2CAP/LE/CFC/BV-30-C<br>L2CAP/LE/CFC/BV-31-C<br>L2CAP/LE/CFC/BI-02-C<br>L2CAP/LE/CID/BI-01-C<br>L2CAP/ECFC/BV-01-C<br>L2CAP/ECFC/BV-02-C<br>L2CAP/ECFC/BV-03-C<br>L2CAP/ECFC/BV-04-C<br>L2CAP/ECFC/BV-06-C<br>L2CAP/ECFC/BV-07-C<br>L2CAP/ECFC/BI-01-C<br>L2CAP/ECFC/BI-02-C<br>L2CAP/ECFC/BV-08-C<br>L2CAP/ECFC/BV-09-C<br>L2CAP/ECFC/BV-10-C<br>L2CAP/ECFC/BV-11-C<br>L2CAP/ECFC/BV-12-C<br>L2CAP/ECFC/BV-13-C<br>L2CAP/ECFC/BV-14-C<br>L2CAP/ECFC/BV-15-C<br>L2CAP/ECFC/BV-16-C<br>L2CAP/ECFC/BV-17-C<br>L2CAP/ECFC/BV-18-C<br>L2CAP/ECFC/BV-19-C<br>L2CAP/ECFC/BV-20-C<br>L2CAP/ECFC/BV-21-C<br>L2CAP/ECFC/BV-22-C<br>L2CAP/ECFC/BV-23-C<br>L2CAP/ECFC/BI-03-C<br>L2CAP/ECFC/BV-24-C<br>L2CAP/ECFC/BV-25-C<br>L2CAP/ECFC/BI-04-C<br>L2CAP/ECFC/BV-26-C<br>L2CAP/ECFC/BV-27-C<br>L2CAP/ECFC/BV-28-C<br>L2CAP/ECFC/BV-29-C<br>L2CAP/ECFC/BV-30-C<br>L2CAP/ECFC/BV-31-C<br>L2CAP/ECFC/BV-32-C<br>L2CAP/ECFC/BV-33-C<br>L2CAP/ECFC/BV-34-C<br>L2CAP/ECFC/BV-35-C<br>L2CAP/ECFC/BV-38-C<br>L2CAP/ECFC/BV-39-C<br>L2CAP/ECFC/BV-41-C<br>L2CAP/ECFC/BI-05-C<br>L2CAP/ECFC/BI-06-C<br>L2CAP/ECFC/BI-07-C<br>L2CAP/ECFC/BI-08-C<br>L2CAP/TIM/BV-03-C</details>

alwa-nordic commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -19,14 +19,15 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
-#define DATA_MTU_INITIAL 128
-#define DATA_MTU 256
-#define DATA_BUF_SIZE BT_L2CAP_SDU_BUF_SIZE(DATA_MTU)
+#define L2CAP_MPS 96
```
Consider setting this to `BT_L2CAP_RX_MTU`, if this is an arbitrary value. `BT_L2CAP_RX_MTU` is the maximum supported by the Host.

sjanc replied:
yeap,

although this is to match current PTS workspace configuration in autopts, we can change it later on but I didn't wanted to touch workspace yet

Thalley replied:
@sjanc Should we consider have a ixit.h file that contains all the configured values for PTS? Then it will be easier to determine if a value is a specific value due to IXIT, or just some arbitrary value, and whether modifying said value requires a change in IXIT as well?

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2972,6 +2975,41 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
 		}
+
+#if defined(CONFIG_BT_L2CAP_SEG_RECV)
+		/**
+		 * If more than one channel is being configured, the MPS field
```
```suggestion
		/* If more than one channel is being configured, the MPS field
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2972,6 +2975,41 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+
+		if (i == 0) {
+			mps = BT_L2CAP_LE_CHAN(chans[i])->rx.mps;
+		}
+
+		if (i > 0) {
```
```suggestion
		} else {
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2972,6 +2975,41 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		}
+
+		if (i > 0) {
+			/* For channels using segment direct API we can only
+			 * decrease MPS and this is not allowed for multiple
+			 * channels. So if any channel is using this MPS cannot
+			 * change.
```
But aren't we here only checking if any previous handled channels use `seg_direct`? 

For example if `chans[1]` does not use it, then it is false here and the `if (mps != BT_L2CAP_LE_CHAN(chans[i])->rx.mps) {` isn't applied, even if `chans[2]` uses it. 

For this check to make sense, don't we need to iterate on all streams first to set the `seg_direct` before we start doing this check?

sjanc replied:
for channels not using seg direct API it is possible to increase MPS (ie if previously it was clipped due to small MTU),  so this check matters only if we step on channel using seg direct

I'll see if those can be more explicit, but initially wanted to keep changes minimal and would multiple loops (but maybe this doesn't hurt)

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (1)</summary>L2CAP   L2CAP/COS/ECFC/BV-04-C          INCONC</details>
<details><summary>Successful tests (92)</summary>L2CAP   L2CAP/COS/CFC/BV-01-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-02-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-03-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-04-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-05-C           PASS<br>L2CAP   L2CAP/COS/ECFC/BV-01-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-02-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-03-C          PASS<br>L2CAP   L2CAP/ECFC/BI-01-C              PASS<br>L2CAP   L2CAP/ECFC/BI-02-C              PASS - ERRATA ES-26416<br>L2CAP   L2CAP/ECFC/BI-03-C              PASS<br>L2CAP   L2CAP/ECFC/BI-04-C              PASS<br>L2CAP   L2CAP/ECFC/BI-05-C              PASS<br>L2CAP   L2CAP/ECFC/BI-06-C              PASS<br>L2CAP   L2CAP/ECFC/BI-07-C              PASS<br>L2CAP   L2CAP/ECFC/BI-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-01-C              PASS<br>L2CAP   L2CAP/ECFC/BV-02-C              PASS<br>L2CAP   L2CAP/ECFC/BV-03-C              PASS<br>L2CAP   L2CAP/ECFC/BV-04-C              PASS<br>L2CAP   L2CAP/ECFC/BV-06-C              PASS<br>L2CAP   L2CAP/ECFC/BV-07-C              PASS<br>L2CAP   L2CAP/ECFC/BV-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-09-C              PASS<br>L2CAP   L2CAP/ECFC/BV-10-C              PASS<br>L2CAP   L2CAP/ECFC/BV-11-C              PASS<br>L2CAP   L2CAP/ECFC/BV-12-C              PASS<br>L2CAP   L2CAP/ECFC/BV-13-C              PASS<br>L2CAP   L2CAP/ECFC/BV-14-C              PASS<br>L2CAP   L2CAP/ECFC/BV-15-C              PASS<br>L2CAP   L2CAP/ECFC/BV-16-C              PASS<br>L2CAP   L2CAP/ECFC/BV-17-C              PASS<br>L2CAP   L2CAP/ECFC/BV-18-C              PASS<br>L2CAP   L2CAP/ECFC/BV-19-C              PASS<br>L2CAP   L2CAP/ECFC/BV-20-C              PASS<br>L2CAP   L2CAP/ECFC/BV-21-C              PASS<br>L2CAP   L2CAP/ECFC/BV-22-C              PASS<br>L2CAP   L2CAP/ECFC/BV-23-C              PASS<br>L2CAP   L2CAP/ECFC/BV-24-C              PASS<br>L2CAP   L2CAP/ECFC/BV-25-C              PASS<br>L2CAP   L2CAP/ECFC/BV-26-C              PASS<br>L2CAP   L2CAP/ECFC/BV-27-C              PASS<br>L2CAP   L2CAP/ECFC/BV-28-C              PASS<br>L2CAP   L2CAP/ECFC/BV-29-C              PASS<br>L2CAP   L2CAP/ECFC/BV-30-C              PASS<br>L2CAP   L2CAP/ECFC/BV-31-C              PASS<br>L2CAP   L2CAP/ECFC/BV-32-C              PASS<br>L2CAP   L2CAP/ECFC/BV-33-C              PASS<br>L2CAP   L2CAP/ECFC/BV-34-C              PASS<br>L2CAP   L2CAP/ECFC/BV-35-C              PASS<br>L2CAP   L2CAP/ECFC/BV-38-C              PASS<br>L2CAP   L2CAP/ECFC/BV-39-C              PASS<br>L2CAP   L2CAP/ECFC/BV-41-C              PASS<br>L2CAP   L2CAP/LE/CFC/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-03-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-04-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-05-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-06-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-07-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-08-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-09-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-10-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-11-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-12-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-13-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-14-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-15-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-16-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-18-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-19-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-20-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-21-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-22-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-23-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-24-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-25-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-26-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-27-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-28-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-29-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-30-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-31-C            PASS<br>L2CAP   L2CAP/LE/CID/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-02-C            PASS<br>L2CAP   L2CAP/LE/REJ/BI-02-C            PASS - ERRATA ES-25764<br>L2CAP   L2CAP/TIM/BV-03-C               PASS</details>

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2973,6 +2975,41 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+				}
+			}
+
+			mps = MAX(mps, BT_L2CAP_LE_CHAN(chans[i])->rx.mps);
```
The MPS should not be the maximum. The MPS in the reconfigure request is the new MPS for all the listed channels.

The right thing to do here is to error out if the wanted MPSs are not all equal.

Note, the MPS of non-seg_direct channels is always `MIN(BT_L2CAP_LE_CHAN(chans[i])->rx.mps, BT_L2CAP_RX_MTU)` and does not ever change.

--------

I think we should be able to add this assert, if it helps:

`!BT_L2CAP_LE_CHAN(chans[i])->chan.ops->recv || BT_L2CAP_LE_CHAN(chans[i])->rx.mps == MIN(BT_L2CAP_LE_CHAN(chans[i])->rx.mps, BT_L2CAP_RX_MTU)`

Here, the MPS should ideally be just equal to `BT_L2CAP_RX_MTU`. But for some reason we complicate this in Zephyr. See the following code comment, that's not citing any source for its claim.

https://github.com/zephyrproject-rtos/zephyr/blob/e7db0f8aff81436ff07f557a5ca8e2daa1789188/subsys/bluetooth/host/l2cap.c#L1226-L1230

alwa-nordic replied:
> and does not ever change.

Actually, that part is wrong, since we do the complicated calculation. It would be right if it just always stayed at `BT_L2CAP_RX_MTU`. Fixed.

sjanc replied:
normally all channels have same MPS and exception are only channels that use segment direct API,    currently on reconfigure such channels will be reset to BT_L2CAP_RX_MTU which might be bigger than what was set by applications.  This MAX() is to ensure that we don't unnecessary clip MPS, but yeah maybe I could just assert if those are not same since  channels with seg direct API are already handled above

BTW I'm not sure why MPS is configurable in segmented direct API channels, instead of simply using same MPS as all other channels.




regarding clipping MPS to MTU...
not sure if there are other IUTs that have this problem, but at least PTS gets triggered by this,  ie if MTU<MPS it may send SDU>MTU in some tests (probably due to hardcoded values),    this means extra work reporting that to PTS folks, testing their patches and broken tests until PTS with fix included is released

alwa-nordic replied:
> BTW I'm not sure why MPS is configurable in segmented direct API channels, instead of simply using same MPS as all other channels.

The reason I made it configurable is that setting the MPS gives application services some control over what one credit is worth. Say the application is able to receive a maximum of 128 bytes of data at the moment. If the MPS is set to 64, it can send 2 credits. If the MPS is set globally to e.g. 256, it can't safely send even one credit.

A global setting is not very flexible when multiple protocols use L2CAP. You may want one MPS for ATT, and another for other uses.

But I am the first to say that I'm not sure how the designers of L2CAP credits envisioned this will be used. It's a awkward system in conjunction with SDUs, which 1. "steals" two bytes of the MPS in the first segment for the SDU length header, and 2. you don't really know how many credits you need to complete an SDU.

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2991,6 +3002,30 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		/* For now simply don't allow MPS changes on channels using
+		 * segment direct API
+		 *
+		 * TODO we could be smarted here
```
```suggestion
		 * TODO we could be smarter here
```

By doing what? A TODO should be a bit more elaborate IMO :)

sjanc replied:
be smarter to allow for MPS change on those channels too :)
in some cases we could allow to reduce MPS (and with some additional tracking increase it)

sjanc commented:
#AutoPTS run zephyr L2CAP

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2991,6 +3002,31 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+
+	for (int j = 0; j < i; j++) {
+		ch = BT_L2CAP_LE_CHAN(chans[j]);
+		/* For now simply don't allow MPS changes on channels using
+		 * segment direct API
```
Is this a behavior change in the stack? 

If so, do we need to update the release notes/migration guide? 
Is this behavior documented in the public API?

alwa-nordic replied:
This public API `bt_l2cap_ecred_chan_reconfigure` does not take an `mps` parameter at all, so there is no good way for the application to provide a new MPS value.

sjanc replied:
this is experimental API and reconfigure is currently not working well for it anyway (is application uses MPS<BT_L2CAP_RX_MTU)

sjanc committed:
Bluetooth: Tester: Unify L2CAP and ACL bufs configuration

This should all be unified so that tests run with similar setup
on all boards.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
#AutoPTS run zephyr L2CAP/COS/ECFC/BV-04-C

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/80302#issuecomment-2437927477, board: nrf52, estimated start time: 16:16:44, test case count: 1, estimated duration: 0:00:23<details><summary>Test cases to be run</summary>L2CAP/COS/ECFC/BV-04-C</details>

sjanc commented:
#AutoPTS run zephyr L2CAP

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
<details><summary>Successful tests (93)</summary>L2CAP   L2CAP/COS/CFC/BV-01-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-02-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-03-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-04-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-05-C           PASS<br>L2CAP   L2CAP/COS/ECFC/BV-01-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-02-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-03-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-04-C          PASS<br>L2CAP   L2CAP/ECFC/BI-01-C              PASS<br>L2CAP   L2CAP/ECFC/BI-02-C              PASS - ERRATA ES-26416<br>L2CAP   L2CAP/ECFC/BI-03-C              PASS<br>L2CAP   L2CAP/ECFC/BI-04-C              PASS<br>L2CAP   L2CAP/ECFC/BI-05-C              PASS<br>L2CAP   L2CAP/ECFC/BI-06-C              PASS<br>L2CAP   L2CAP/ECFC/BI-07-C              PASS<br>L2CAP   L2CAP/ECFC/BI-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-01-C              PASS<br>L2CAP   L2CAP/ECFC/BV-02-C              PASS<br>L2CAP   L2CAP/ECFC/BV-03-C              PASS<br>L2CAP   L2CAP/ECFC/BV-04-C              PASS<br>L2CAP   L2CAP/ECFC/BV-06-C              PASS<br>L2CAP   L2CAP/ECFC/BV-07-C              PASS<br>L2CAP   L2CAP/ECFC/BV-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-09-C              PASS<br>L2CAP   L2CAP/ECFC/BV-10-C              PASS<br>L2CAP   L2CAP/ECFC/BV-11-C              PASS<br>L2CAP   L2CAP/ECFC/BV-12-C              PASS<br>L2CAP   L2CAP/ECFC/BV-13-C              PASS<br>L2CAP   L2CAP/ECFC/BV-14-C              PASS<br>L2CAP   L2CAP/ECFC/BV-15-C              PASS<br>L2CAP   L2CAP/ECFC/BV-16-C              PASS<br>L2CAP   L2CAP/ECFC/BV-17-C              PASS<br>L2CAP   L2CAP/ECFC/BV-18-C              PASS<br>L2CAP   L2CAP/ECFC/BV-19-C              PASS<br>L2CAP   L2CAP/ECFC/BV-20-C              PASS<br>L2CAP   L2CAP/ECFC/BV-21-C              PASS<br>L2CAP   L2CAP/ECFC/BV-22-C              PASS<br>L2CAP   L2CAP/ECFC/BV-23-C              PASS<br>L2CAP   L2CAP/ECFC/BV-24-C              PASS<br>L2CAP   L2CAP/ECFC/BV-25-C              PASS<br>L2CAP   L2CAP/ECFC/BV-26-C              PASS<br>L2CAP   L2CAP/ECFC/BV-27-C              PASS<br>L2CAP   L2CAP/ECFC/BV-28-C              PASS<br>L2CAP   L2CAP/ECFC/BV-29-C              PASS<br>L2CAP   L2CAP/ECFC/BV-30-C              PASS<br>L2CAP   L2CAP/ECFC/BV-31-C              PASS<br>L2CAP   L2CAP/ECFC/BV-32-C              PASS<br>L2CAP   L2CAP/ECFC/BV-33-C              PASS<br>L2CAP   L2CAP/ECFC/BV-34-C              PASS<br>L2CAP   L2CAP/ECFC/BV-35-C              PASS<br>L2CAP   L2CAP/ECFC/BV-38-C              PASS<br>L2CAP   L2CAP/ECFC/BV-39-C              PASS<br>L2CAP   L2CAP/ECFC/BV-41-C              PASS<br>L2CAP   L2CAP/LE/CFC/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-03-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-04-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-05-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-06-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-07-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-08-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-09-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-10-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-11-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-12-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-13-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-14-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-15-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-16-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-18-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-19-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-20-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-21-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-22-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-23-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-24-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-25-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-26-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-27-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-28-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-29-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-30-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-31-C            PASS<br>L2CAP   L2CAP/LE/CID/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-02-C            PASS<br>L2CAP   L2CAP/LE/REJ/BI-02-C            PASS - ERRATA ES-25764<br>L2CAP   L2CAP/TIM/BV-03-C               PASS</details>

alwa-nordic reviewed (changes requested):
The proposed `bt_l2cap_ecred_chan_reconfigure` will copy the MTU from a `seg_recv` channel in the list to the other channels. This is complex and hard to document.

Worse, this can result in a decrease of the MPS which is illegal when the reconfigure applies to multiple channels.

Possible solution: I think for simplicity, we can forbid using `bt_l2cap_ecred_chan_reconfigure` with `seg_recv` channels, at least for now.

And we could add a separate "`bt_l2cap_ecred_chan_seg_recv_reconfigure`" that takes both MTU and MPS as input.

sjanc commented:
Hi,

Well, current bt_l2cap_ecred_chan_reconfigure() has this issue with MPS decrease anyway,  I can extend check for this

I'm fine with new API too that would allow for explicit MPS setting and be allowed only for seg direct channels

alwa-nordic commented:
> Well, current bt_l2cap_ecred_chan_reconfigure() has this issue with MPS decrease anyway, I can extend check for this

Thanks, an explicit check is a good idea anyway. But, it happens that there is no issue for the reason that the MTU of a channel cannot decrease, by spec. This is verified in the function. And the calculated value of MPS grows or stays constant when MTU grows. So we can actually assert this.

sjanc commented:
runtime check needs to be there since seg direct channel may have smaller mps than stack default

alwa-nordic commented:
Now the proposed bt_l2cap_ecred_chan_reconfigure will copy the MPS from the seg_recv channels in the list to the other channels. This is non-obvious so the documentation of `bt_l2cap_ecred_chan_reconfigure` has to be updated to reflect this.

But I would prefer if we find a way to keep the API simple. I suggest one of:

  - Returning an error if `bt_l2cap_ecred_chan_reconfigure` is used with `seg_recv`.
  - Specify and document that `bt_l2cap_ecred_chan_reconfigure` has an implicit parameter `mps` set to `MIN(mtu + BT_L2CAP_SDU_HDR_SIZE, BT_L2CAP_RX_MTU)`. It behaves as-if this was an explicit parameter from the caller and simply errors out if this choice of MPS is not valid.

sjanc commented:
> Now the proposed bt_l2cap_ecred_chan_reconfigure will copy the MPS from the seg_recv channels in the list to the other channels. This is non-obvious so the documentation of `bt_l2cap_ecred_chan_reconfigure` has to be updated to reflect this.
> 
> But I would prefer if we find a way to keep the API simple. I suggest one of:
> 
> * Returning an error if `bt_l2cap_ecred_chan_reconfigure` is used with `seg_recv`.

We need this functionality for upper tester to pass qualification tests 

> * Specify and document that `bt_l2cap_ecred_chan_reconfigure` has an implicit parameter `mps` set to `MIN(mtu + BT_L2CAP_SDU_HDR_SIZE, BT_L2CAP_RX_MTU)`. It behaves as-if this was an explicit parameter from the caller and simply errors out if this choice of MPS is not valid.

I agree that updating docs should be sufficient to not confuse user

alwa-nordic commented:
Good. And then when someone need explicit control of the MPS in the reconfigure, we will either make the `mps` parameter explicit, or we will create a second reconfigure API that also takes the MPS as parameter.

sjanc commented:
superseded by https://github.com/zephyrproject-rtos/zephyr/pull/80911
